File: Scala/src/main/scala/TextbookTrack/Chapter03/BA3I.scala
package TextbookTrack.Chapter03

object BA3I {
  import scala.annotation.tailrec
  import scala.collection.mutable.{Map => MutableMap}

  class EulerianGraph[T](adjacencyList: Map[T, List[T]])(implicit ord: Ordering[T]) {
    import EulerianGraph.{extendCycle, findNodeWithUnusedEdges, shiftCycle}

    def findEulerianCycle(): List[T] = {
      @tailrec
      def loop(edges: Map[T, List[T]], cycle: List[T]): List[T] = {
        if (edges.isEmpty) cycle.reverse
        else {
          val (nextStartNode, ix): (T, Int) = findNodeWithUnusedEdges(edges, cycle)
          val shiftedCycle: List[T] = shiftCycle(cycle, ix, nextStartNode)
          val (remainingEdges, extendedCycle): (Map[T, List[T]], List[T]) =
            extendCycle(edges, nextStartNode, shiftedCycle)
          loop(remainingEdges, extendedCycle)
        }
      }

      val startNode: T = adjacencyList.keysIterator.min
      loop(adjacencyList, List(startNode))
    }
  }

  object EulerianGraph {
    private def findNodeWithUnusedEdges[T](adjacencyList: Map[T, List[T]], cycle: List[T]): (T, Int) =
      cycle.iterator.zipWithIndex.find { case (node, _) => adjacencyList.contains(node) } match {
        case None => throw new Exception("The cycle must contain nodes with unused outgoing edges.")
        case Some((node, ix)) => (node, ix)
      }

    private def shiftCycle[T](cycle: List[T], ix: Int, startNode: T): List[T] = {
      val (first, second): (List[T], List[T]) = cycle.splitAt(ix + 1)
      startNode :: second ::: first.tail
    }

    private def extendCycle[T](adjacencyList: Map[T, List[T]],
                               startNode: T,
                               initialCycle: List[T]): (Map[T, List[T]], List[T]) = {
      @tailrec
      def loop(edges: Map[T, List[T]], cycle: List[T], currentNode: T): (Map[T, List[T]], List[T]) = {
        if (!edges.contains(currentNode)) (edges, cycle)
        else {
          edges.getOrElse(currentNode, Nil) match {
            case Nil => throw new Exception(s"The node $currentNode should have been removed earlier.")
            case nextNode :: rest =>
              val updatedEdges: Map[T, List[T]] =
                if (rest.isEmpty) edges.removed(currentNode) else edges.updated(currentNode, rest)
              loop(updatedEdges, nextNode :: cycle, nextNode)
          }
        }
      }

      loop(adjacencyList, initialCycle, startNode)
    }
  }

  private def cartesianProduct[T](xs: Seq[T], n: Int): Iterator[List[T]] =
    (0 until n).foldLeft(Iterator(List.empty[T]))((acc, _) => acc.flatMap(ys => xs.map(_ :: ys)))

  private def createAdjacencyListFromBinaryStrings(k: Int): Map[String, List[String]] = {
    val binaryKMers: Iterator[String] = cartesianProduct(List('0', '1'), k).map(_.mkString)
    val adjacencyList: MutableMap[String, List[String]] = MutableMap.empty
    binaryKMers.foreach{ kMer =>
      val prefix: String = kMer.dropRight(1)
      val suffix: String = kMer.drop(1)
      val neighbours: List[String] = adjacencyList.getOrElse(prefix, Nil)
      adjacencyList.update(prefix, suffix :: neighbours)
    }
    adjacencyList.toMap
  }

  private def calcStringSpelledByAGenomePath(kMers: List[String]): String = kMers match {
    case Nil => ""
    case x :: xs => (x.toList ::: xs.map(_.last)).mkString
  }

  def findUniversalCircularBinaryString(k: Int): String = {
    if (k <= 0) ""
    else if (k == 1) "01"
    else {
      val adjacencyList: Map[String, List[String]] = createAdjacencyListFromBinaryStrings(k)
      val graph = new EulerianGraph(adjacencyList)
      val cycle: List[String] = graph.findEulerianCycle()
      val universalString: String = calcStringSpelledByAGenomePath(cycle)
      universalString.dropRight(k - 1)
    }
  }

  def main(args: Array[String]): Unit = {
    val reader: Iterator[String] = scala.io.Source.stdin.getLines()
    val k: Int = reader.next().toInt
    val result: String = findUniversalCircularBinaryString(k)
    println(result)
  }
}

File: Python/textbook_track/chapter03/ba3i.py
# Find a k-Universal Circular String
import itertools as it
import sys

from collections import defaultdict


def create_adjacency_list_from_binary_strings(k):
    d = defaultdict(list)
    for k_mer in map(''.join, it.product('01', repeat=k)):
        prefix = k_mer[:-1]
        suffix = k_mer[1:]
        d[prefix].append(suffix)
    return dict(d)


def find_node_with_unused_edges_on_cycle(adjacency_list, cycle):
    for ix, node in enumerate(cycle):
        if node in adjacency_list:
            return ix, node


def shift_cycle(cycle, ix):
    next_start_node = cycle[ix]
    _ = cycle.pop(-1)
    new_cycle = cycle[ix:] + cycle[:ix]
    new_cycle.append(next_start_node)
    return new_cycle


def extend_cycle(adjacency_list, start_node, cycle=None):
    if cycle is None:
        cycle = [start_node]

    current_node = start_node
    while current_node in adjacency_list:
        neighbours = adjacency_list[current_node]
        next_node = neighbours.pop(-1)
        if not neighbours:
            _ = adjacency_list.pop(current_node)

        cycle.append(next_node)
        current_node = next_node
    return cycle


def find_eulerian_cycle(adjacency_list):
    start_node = min(adjacency_list)
    cycle = extend_cycle(adjacency_list, start_node)
    while adjacency_list:
        ix, new_start_node = find_node_with_unused_edges_on_cycle(adjacency_list, cycle)
        shifted_cycle = shift_cycle(cycle, ix)
        cycle = extend_cycle(adjacency_list, new_start_node, shifted_cycle)
    return cycle


def calc_string_spelled_by_a_genome_path(k_mers):
    text = list(k_mers[0])
    for k_mer in k_mers[1:]:
        text.append(k_mer[-1])
    return ''.join(text)


def find_universal_circular_binary_string(k):
    if k <= 0:
        return ''

    if k == 1:
        return '01'

    adjacency_list = create_adjacency_list_from_binary_strings(k)
    cycle = find_eulerian_cycle(adjacency_list)
    universal_string = calc_string_spelled_by_a_genome_path(cycle)
    return universal_string[:-(k-1)]


def main():
    reader = sys.stdin
    k = int(next(reader))
    result = find_universal_circular_binary_string(k)
    print(result)


if __name__ == '__main__':
    main()
